django_celery_beat/schedulers.py

```diff
@@ -248,18 +256,36 @@ def sync(self):
             self._dirty |= _tried
             logger.exception('Database error while sync: %r', exc)
 
-    def update_from_dict(self, mapping):
+    def update_from_dict(self, mapping, origin_key=None):
         s = {}
         for name, entry_fields in items(mapping):
             try:
                 entry = self.Entry.from_entry(name,
                                               app=self.app,
+                                              origin_key=origin_key,
                                               **entry_fields)
                 if entry.model.enabled:
                     s[name] = entry
 
             except Exception as exc:
                 logger.error(ADD_ENTRY_ERROR, name, exc, entry_fields)
+        if origin_key:
+            # delete database-persisted periodic tasks that
+            # are no longer declared in source code
+            existing_task_instances = PeriodicTask.objects.filter(
+                origin_key=origin_key
+            )
+            existing_tasks = set(
+                map(lambda x: x.name, existing_task_instances)
+            )
```

Comment:
Could replace a lot of logic here with just this I think: 

    tasks_to_purge = PeriodicTask.objects.filter(
        origin_key=origin_key).exclude(name__in=mapping.keys())
    if tasks_to_purge:
        tasks_to_purge.delete()
        log.warn(...)